### ISSUE TITLE
Build backend tracking issue

### ISSUE THREAD

**Issue by konstin** [astral-sh/uv#8779]
uv should provide its own build backend. This issue tracks this work, it will become more granular as more things are implemented.

- [x] Support all pyproject.toml METADATA
- [x] Support PEP 639
- [x] Build basic wheels
- [ ] Build basic source distributions
  - [ ] Include files
  - [ ] Static metadata
- [ ] Test source tree -> source dist -> wheels
- [ ] Check inclusions and files, similar to https://github.com/astral-sh/uv/issues/8774 and https://github.com/astral-sh/uv/issues/8641
- [ ] Documentation
- [ ] Patterns for inclusions and exclusions: This is one of the most complex parts and will evolve in iterations
- [ ] Plan for future features: Native code and plugins
- [ ] Preview release
- [ ] Stable release

**Comment by hauntsaninja** [comment]
Check MANIFEST.in could also be nice! https://github.com/mgedmin/check-manifest

**Comment by cthoyt** [comment]
@konstin I was able to get my first builds working using this on the 0.5.5 release (also w/ tox and tox-uv)! There were a few tricks and rough edges I had to get around to make it work.

Are you interested in feedback at this point? I totally understand if not - you are probably aware of a lot of things already and I don't want to bog you down.

If so, what's the preferred mechanism? I am happy to open up issues with minimum reproducibility scripts where appropriate.

**Comment by konstin** [comment]
With #9621, everything that's needed for alpha testing has now landed on main. I've updated the original post with some basic documentation.

tl;dr: Set `UV_PREVIEW=1` and add:

```
[build-system]
requires = ["uv>=0.5,<0.6"]
build-backend = "uv"
```

---

The functionality for the basic build backend and its integration has landed (all behind preview), the biggest item to still change is the include/exclude syntax, here we want to sync up with red knot and optimize towards typical project usage without loss of generality (see also the ecosystem review in the original post).

Currently not part of the design is generating files or any part of build step, such as cython compilation, code generation from schema, but also compiling native modules in C, C++ or Rust. The includes allow including extra files in the source dist for a future compilation or plugin step.

---

@cthoyt It's great to have early adopters! If you have a bug or specific feature request, just open a new issue, minimal reproductions are of course always greatly appreciated :) For general discussion on what the build backend should (or shouldn't) do and other design discussions, I'd use #3957, while I'll post general updates in this issue. Please share your feedback, i love hearing from users!

**Comment by uwu-420** [comment]
Really excited to see progress on uv's own build backend <3

I know this is currently out of scope, but I'd be interested if there are plans to eventually extend the build backend to handle building for multiple python versions in one go. Currently one would have to use tools like tox/no, cibuildwheel or a custom build script but I think it would be wonderful to only run `uv build` and call it a day. Of course with some previous configuration e.g. specifying the desired target versions.

Or am I missing the point what a build backend should handle or if there is already an idiomatic way to do this?

**Comment by konstin** [comment]
In the latest release, we've split the build backend out into a separate `uv_build` package. The `uv` package itself doesn't contain the build backend anymore, I've updated the documentation in #8779 accordingly.

**Comment by cthoyt** [comment]
@konstin thank you for the hard work! It seems that it's not so easy to just update the configuration, for example in my PR https://github.com/biopragmatics/pyobo/pull/364, I'm getting the error

```
Error: Unknown subcommand: build-backend (cli: ArgsOs { inner: ["/Users/cthoyt/dev/pyobo/.tox/.pkg/bin/uv-build", "build-backend", "build-sdist", "/Users/cthoyt/dev/pyobo/.tox/.pkg/dist"] })
Traceback (most recent call last):
  File "/Users/cthoyt/Library/Application Support/uv/tools/tox/lib/python3.13/site-packages/pyproject_api/_backend.py", line 93, in run
    outcome = backend_proxy(parsed_message["cmd"], **parsed_message["kwargs"])
  File "/Users/cthoyt/Library/Application Support/uv/tools/tox/lib/python3.13/site-packages/pyproject_api/_backend.py", line 34, in __call__
    return getattr(on_object, name)(*args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/Users/cthoyt/dev/pyobo/.tox/.pkg/lib/python3.13/site-packages/uv_build/__init__.py", line 64, in build_sdist
    return call(args, config_settings)
  File "/Users/cthoyt/dev/pyobo/.tox/.pkg/lib/python3.13/site-packages/uv_build/__init__.py", line 48, in call
    sys.exit(result.returncode)
    ~~~~~~~~^^^^^^^^^^^^^^^^^^^
SystemExit: 1
Backend: run command build_sdist with args {'sdist_directory': '/Users/cthoyt/dev/pyobo/.tox/.pkg/dist', 'config_settings': None}
Backend: Wrote response {'code': 1, 'exc_type': 'SystemExit', 'exc_msg': '1'} to /var/folders/6k/r2gmmc213vl6h7bv20x624xw0000gn/T/pep517_build_sdist-uyoantdq.json
```

**Comment by konstin** [comment]
Sorry, the package had the old Python shim still! Fixed in https://github.com/astral-sh/uv/pull/12058.

**Comment by cthoyt** [comment]
> Sorry, the package had the old Python shim still! Fixed in [#12058](https://github.com/astral-sh/uv/pull/12058).

woohoo! sorry I also opened a dedicated bug report with a MRE in #12059 before I saw your comment. Hopefully that can help testing 🚀

**Comment by konstin** [comment]
We considered a number of different ways of configuring the build backend (see also https://github.com/astral-sh/uv/pull/11446#issuecomment-2689508497). In the end, we stylized the package name in the documentation as `uv_build`, so that `build-system.requires` and `build-sytem.build-backend` use the same separator between `uv` and `build`. We want to avoid confusing users about why both `requires = ["uv_build>=0.6,<0.7"]` and `requires = ["uv-build>=0.6,<0.7"]` are allowed, but then `build-backend = "uv_build"` works while `build-backend = "uv-build"` fails with a `ModuleNotFoundError` by using the underscore everywhere. You are of course free to use the normalized version with the underscore. PyPI also implements the normalization, so https://pypi.org/project/uv_build/ redirects to https://pypi.org/project/uv-build/.

```toml
[build-system]
requires = ["uv_build>=0.6,<0.7"]
build-backend = "uv_build"
```

**Comment by konstin** [comment]
To avoid splitting the conversation between two threads, I'll lock this issue in favor of #3957. I'll post update the top level post and post updates here, but all the discussion (that is not in separate issues) should happen in #3957.

**Comment by konstin** [comment]
Update from v0.7.3: The uv build backend now has both configuration and references documentation and is the default in preview (https://github.com/astral-sh/uv/pull/12804). Builds should now be reproducible between machines of the same platform and in most cases also independent of the  build platform (https://github.com/astral-sh/uv/pull/13171). We also added a way to escape characters in globs, extending over the PEP 639 syntax we previously used (https://github.com/astral-sh/uv/pull/13313).

**Comment by konstin** [comment]
In 0.7.13, we added support for namespaces. For packages with a single root module, you can set `module-name` for `src/cloud/sdk/db/__init__.py`:

```
[tool.uv.build-backend]
module-name = "cloud.sdk.db"
```

You can also set `namespace = true` for complex namespace packages with more than one root module.

The build backend now also supports stubs packages, including for namespace packages:

```
[tool.uv.build-backend]
module-name = "cloud-stubs.sdk.db"
```

Check the documentation for details: https://docs.astral.sh/uv/concepts/build-backend/

The core of the build backend should be feature complete with those additions and unless we find problems with the configuration, ready for stabilization. Stabilization here doesn't mean that we won't add more feature, but that all core features are present and ready for general use.

**Comment by konstin** [comment]
The uv build backend is now stable - ready for production - as of 0.7.19! See the docs on how to gets started: https://docs.astral.sh/uv/concepts/build-backend/

We will follow up with with problems and feature requests (please open issues!). We don't cover everything a uv build system could do in this release, you can feature requests under the [issues with the build backend tag](https://github.com/astral-sh/uv/issues?q=state%3Aopen%20label%3A%22build-backend%22), and, as usual, find new features in our release notes.

One change that's in the pipeline that I can already announce is that we'll make the uv build backend the default in `uv init` (which is currently still hatchling as it's a behavior change).
